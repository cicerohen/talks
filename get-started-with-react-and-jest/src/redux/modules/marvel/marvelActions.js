import MODULE_NAME from './marvelConstants';

export const FETCH_CHARACTERS_REQUESTED = `${MODULE_NAME}/FETCH_CHARACTERS_REQUESTED`;
export const FETCH_CHARACTERS_COMPLETED = `${MODULE_NAME}/FETCH_CHARACTERS_COMPLETED`;
export const FETCH_CHARACTERS_FAILED = `${MODULE_NAME}/FETCH_CHARACTERS_FAILED`;

export const fetchCharacters = () => ({
  type: FETCH_CHARACTERS_REQUESTED,
});

export const completeFetchCharacters = data => ({
  type: FETCH_CHARACTERS_COMPLETED,
  payload: {
    data,
  },
});

export const failFetchCharacters = error => ({
  type: FETCH_CHARACTERS_FAILED,
  payload: {
    error,
  },
});
