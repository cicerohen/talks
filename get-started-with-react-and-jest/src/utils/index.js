import { Iterable, fromJS } from 'immutable';
import { createBrowserHistory } from 'history';

export const asImmutable = obj => (Iterable.isIterable(obj) ? obj : fromJS(obj));
export const emptyMap = asImmutable({});

export const getHistory = () => {
    window.browserHistory = window.browserHistory || createBrowserHistory();
    return window.browserHistory;
};

export const isDevEnv = () => {
  const nodeEnv = process.env.NODE_ENV;
  return nodeEnv !== 'production' && nodeEnv !== 'test';
};
