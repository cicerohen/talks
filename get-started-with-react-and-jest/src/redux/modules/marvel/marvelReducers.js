import { asImmutable } from '../../../utils';
import * as actions from './marvelActions';

const initialState = asImmutable({
  isFetchingCharacters: false,
  fetchError: {},
  characters: [],
});

export const reducer = (state = initialState, action) => {
    switch (action.type) {
      case actions.FETCH_CHARACTERS_REQUESTED:
        return state.set('isFetchingCharacters', true);
      case actions.FETCH_CHARACTERS_COMPLETED:
        return state.merge({
          isFetchingCharacters: false,
          characters: action.payload.data,
        });
      case actions.FETCH_CHARACTERS_FAILED:
        return state.merge({
          isFetchingCharacters: false,
          fetchError: action.payload.error,
        });
      default:
        return state;
    }
};

