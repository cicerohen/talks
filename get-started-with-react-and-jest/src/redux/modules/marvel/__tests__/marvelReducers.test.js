import { reducer } from '../marvelReducers';
import * as actions from '../marvelActions';
import { asImmutable } from '../../../../utils';

const initialState = asImmutable({
  isFetchingCharacters: false,
  fetchError: {},
  characters: [],
});

describe('marvel/marvelReducers.js Tests', () => {
  it('should save state correctly with fetchCharacters action', (done) => {
    expect(reducer(initialState, actions.fetchCharacters())).toMatchSnapshot();
    done();
  });
  
  it('should save state correctly with completeFetchCharacters action', (done) => {
    const data = [{
      id: 1,
      name: 'character name',
    }];
    expect(reducer(initialState, actions.completeFetchCharacters(data))).toMatchSnapshot();
    done();
  });

  it('should save state correctly with failFetchCharacters action', (done) => {
    const error = {
      type: 'Error',
    };
    expect(reducer(initialState, actions.failFetchCharacters(error))).toMatchSnapshot();
    done();
  })
});