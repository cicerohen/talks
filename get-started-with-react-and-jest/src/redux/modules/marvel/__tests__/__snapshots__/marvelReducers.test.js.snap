// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`marvel/marvelReducers.js Tests should save state correctly with completeFetchCharacters action 1`] = `
Immutable.Map {
  "isFetchingCharacters": false,
  "fetchError": Immutable.Map {},
  "characters": Immutable.List [
    Immutable.Map {
      "id": 1,
      "name": "character name",
    },
  ],
}
`;

exports[`marvel/marvelReducers.js Tests should save state correctly with failFetchCharacters action 1`] = `
Immutable.Map {
  "isFetchingCharacters": false,
  "fetchError": Immutable.Map {
    "type": "Error",
  },
  "characters": Immutable.List [],
}
`;

exports[`marvel/marvelReducers.js Tests should save state correctly with fetchCharacters action 1`] = `
Immutable.Map {
  "isFetchingCharacters": true,
  "fetchError": Immutable.Map {},
  "characters": Immutable.List [],
}
`;
